src/contrib/inference/dynamichmc.jl

```diff
@@ -0,0 +1,94 @@
+###
+### DynamicHMC backend - https://github.com/tpapp/DynamicHMC.jl
+###
+struct DynamicNUTS{AD, space} <: Hamiltonian{AD} end
+
+"""
+    DynamicNUTS()
+
+Dynamic No U-Turn Sampling algorithm provided by the DynamicHMC package.
+To use it, make sure you have the DynamicHMC package installed.
+
+"""
+DynamicNUTS(args...) = DynamicNUTS{ADBackend()}(args...)
+DynamicNUTS{AD}() where AD = DynamicNUTS{AD, ()}()
+function DynamicNUTS{AD}(space::Symbol...) where AD
+    DynamicNUTS{AD, space}()
+end
+
+mutable struct DynamicNUTSState <: AbstractSamplerState
+    vi::VarInfo
+    draws::Vector
+end
+
+getspace(::Type{<:DynamicNUTS{<:Any, space}}) where {space} = space
+getspace(alg::DynamicNUTS{<:Any, space}) where {space} = space
+
+function sample_init!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:DynamicNUTS},
+    N::Integer;
+    kwargs...
+)
+    # Set up lp function.
+    function _lp(x)
+        value, deriv = gradient_logp(x, spl.state.vi, model, spl)
+        return ValueGradient(value, deriv)
+    end
+
+    runmodel!(model, spl.state.vi, SampleFromUniform())
+
+    if spl.selector.tag == :default
+        link!(spl.state.vi, spl)
+        runmodel!(model, spl.state.vi, spl)
+    end
+
+    spl.state.draws, _ = NUTS_init_tune_mcmc(
+        FunctionLogDensity(
+            length(spl.state.vi[spl]),
+            _lp
+        ),
+        N
+    )
+end
+
+function sample_end!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:DynamicNUTS},
+    N::Integer;
+    kwargs...
+)
+    runmodel!(model, spl.state.vi, SampleFromUniform())
+
+    if spl.selector.tag == :default
+        link!(spl.state.vi, spl)
+        runmodel!(model, spl.state.vi, spl)
+    end   
+end
+
+function step!(
```

Comment:
I'm a bit worried about this. Is it the case user can use DHMC in Gibbs as well but in fact it will be buggy?

Reply:
Also might give us performance overhead.

Reply:
We could do the same thing as we did for AHMC (the `steps!`)

Reply:
The correct way to do this is to go into DynamicHMC and integrate with their internal API to use the `step!` function, but DynamicHMC is currently in a state of flux. This code is mostly a stopgap to get DynamicHMC hobbling until we can revisit it later.

I am reluctant to do a `steps!` function because that gets us into a very specialized area that 1) complicates the interface and 2) probably won't really impact performance that much. I suppose though that is a design. As a side note, I've removed the AHMC `steps!` function in this PR and performance appears to be unchanged.

Ideally AHMC should be on a similar interface, and it should extend `sample_init!`, `sample_end!`, and `step!`. The `steps!` function really only just wraps `AHMC.sample`, which could implement the same interface, and Turing would just sit on top of AHMC's extended functions.

Reply:
To answer your first question, yes, you can use this with Gibbs, but it'll give you terrible results. It's possible to intergrate the entire sampler with Gibbs by going back and adapting DynamicHMC for people who want to use it, but I think everyone should be using AHMC anyways because it's pretty spectacular.

Reply:
Thanks for the response. I agree that it should not give much performance differences as inside AHMC they are a loop anyway.